api/qrcodegenerator.proto

```diff
@@ -90,3 +90,52 @@ service EndDeviceQRCodeGenerator {
     };
   };
 }
+
+// EndDeviceOnboardingData contains fields necessary for onboarding a LoRaWAN End Device.
+message EndDeviceOnboardingData {
+	bytes join_eui = 1 [(gogoproto.customtype) = "go.thethings.network/lorawan-stack/v3/pkg/types.EUI64"];
+  bytes dev_eui = 2 [(gogoproto.customtype) = "go.thethings.network/lorawan-stack/v3/pkg/types.EUI64"];
+  // claim_authentication_code ids referred to as;
+  // - Owner Token in LoRa Alliance TR005 specification.
+  // - Device Validation Code in the LoRa Alliance TR005 draft 2 and draft 3 specs.
+  string claim_authentication_code = 3;
+  bytes vendor_id = 4;
+  // model_id is referred to as VendorProfileID in LoRa Alliance TR005 specification.
+  bytes model_id = 5;
+  string checksum = 6;
+  string serial_number = 7;
+  string proprietary = 8;
+}
+
+message EntityOnboardingData {
+  // Identifier of the format used to successfully parse the QR code data.
+  string format_id = 1;
+  oneof data {
+    option (validate.required) = true;
+    EndDeviceOnboardingData end_device_onboarding_data = 2;
+  }
+}
+
+message ParseQRCodeRequest {
+  // QR code format identifier. Enumerate available formats with 
+  // - End Devices: rpc `ListFormats` in the EndDeviceQRCodeGenerator service.
+  // If this field is not specified, the server will attempt to parse the data with each known format.
```

Comment:
How are these things related exactly? This is a generic service, right? What is the namespace of these formats?

Why allow one format? Why not multiple?

Why not limit results to entities, like end devices or gateways (or both)?

Do we need this at all?

Reply:
So the idea is that if the user already knows the format, then we can parse it directly with that format and we don't have to trial and error with all the formats. 

> Why allow one format? Why not multiple?

When a QR code is generated, it's generated for one format right?. If that format is known, then we use that to parse it. Why would we need to allow multiple formats?

> This is a generic service, right?

Indeed. That's why I added the line 
```
- End Devices: rpc `ListFormats` in the EndDeviceQRCodeGenerator service.
```
If we add gateway QR Codes, then I suppose we'll also add a new `GatewayQRCodeGenerator` which would be option 2 in this description. Ex:
```
- End Devices: rpc `ListFormats` in the EndDeviceQRCodeGenerator service.
- Gateways: rpc `ListFormats` in the GatewayQRCodeGenerator service.
```

> Why not limit results to entities, like end devices or gateways (or both)? 
> What is the namespace of these formats?

I think these are related questions.

I suppose we could add an `entity_type` field to the request and users can choose `enddevices`, `gateways` or something else we support later. If the user then provides the `format_id` (ex: We know that it's TR005), then we use that or we loop through each possible case within the formats supported for the entity.  

Do you see a case where the client (ex: Console) reads an unspecified QR code and we need to determine the entity also?

Reply:
> If we add gateway QR Codes, then I suppose we'll also add a new `GatewayQRCodeGenerator` which would be option 2 in this description. Ex:
> 
> ```
> - End Devices: rpc `ListFormats` in the EndDeviceQRCodeGenerator service.
> - Gateways: rpc `ListFormats` in the GatewayQRCodeGenerator service.
> ```

I see, but that's what I meant with the namespace. TR005 is now about end devices, but TR042 could be about end devices and gateways. The format IDs are unique within `EndDeviceQRCodeGenerator` but not in `QRCodeParser`.

> I don't know if we need to be more generic than that. Do you see a case where the client (ex: Console) reads an unspecified QR code and we need to determine the entity also?

Not really, actually. Maybe we should keep this really simple and make a service specifically for end devices and add gateways later.

Reply:
Ok so this turns out pretty neat with a new RPC to the existing `EndDeviceQRCodeGenerator`.